src/lib/OS/MemRegion.cc

```diff
@@ -83,38 +77,60 @@ uint64_t MemRegion::updateBrkRegion(uint64_t newBrk) {
   return brk_;
 }
 
-uint64_t MemRegion::addVma(VMA* vma, uint64_t startAddr) {
-  VMA* curr = vm_head_;
-  size_t size = vma->vmSize_;
+uint64_t MemRegion::addVma(VMA vma, uint64_t startAddr) {
+  size_t size = vma.vmSize_;
+  auto last = std::prev(VMAlist_->end(), 1);
+  bool allocated = false;
   // When starAddr is not 0, search for an available address range
   // that can hold the new VMA with a starting address that is greater than or
   // equal to the specified startAddr. The following algorithm retrieves the
   // last existing VMA object before the address range so that the new VMA can
   // be linked between two existing VMAs. If no available address range is
   // found, then the new VMA is allocated at the end of the VMA list.
-  if (startAddr && vm_size_ > 0) {
-    while (curr->vmNext_ != nullptr) {
-      if (curr->vmEnd_ <= startAddr &&
-          curr->vmNext_->vmStart_ >= (startAddr + size)) {
-        break;
+  auto itr = VMAlist_->begin();
+  if (VMAlist_->size() > 0) {
+    if (startAddr) {
+      uint64_t space = itr->vmStart_ - startAddr;
+      // Check if the result of subtraction is negative, resulting in a wrapped
+      // unsigned positive value.
+      bool negativeWrapAround = itr->vmStart_ < startAddr;
+      if (!negativeWrapAround && space >= size) {
+        vma.vmStart_ = startAddr;
+        vma.vmEnd_ = startAddr + size;
+        VMAlist_->insert(VMAlist_->begin(), vma);
+        return vma.vmStart_;
+      }
+      for (itr = VMAlist_->begin(); itr != last; itr++) {
+        auto next = std::next(itr, 1);
+        if (itr->vmEnd_ <= startAddr && next->vmStart_ >= (startAddr + size)) {
+          break;
+        }
       }
-      curr = curr->vmNext_;
     }
-  }
+    // If enough space is available between mmapStart and head of VMA list,
+    // allocate the new VMA as the new head of the list.
+    else if (VMAlist_->front().vmStart_ - mmapStart_ >= size) {
+      vma.vmStart_ = mmapStart_;
+      vma.vmEnd_ = mmapStart_ + size;
+      VMAlist_->insert(VMAlist_->begin(), vma);
+      return mmapStart_;
+    }
 
-  bool allocated = false;
-  // If the VMA list has multiple VMAs then starting from curr (VMA) check if
-  // the new VMA can be allocated between two existing ones.
-  while (curr != nullptr && curr->vmNext_ != nullptr) {
-    if (curr->vmNext_->vmStart_ - curr->vmEnd_ >= size) {
-      vma->vmStart_ = curr->vmEnd_;
-      vma->vmNext_ = curr->vmNext_;
-      vma->vmEnd_ = curr->vmEnd_ + size;
-      curr->vmNext_ = vma;
-      allocated = true;
-      break;
+    // If the VMA list has multiple VMAs then starting from curr (VMA) check if
+    // the new VMA can be allocated between two existing ones.
+    while (itr != VMAlist_->end() && itr != last) {
```

Comment:
This loop triggers on both conditions i.e when startAddr is zero and non-zero. This loop looks for space between existing VMAs such that the new VMA can be allocated between two existing ones.

The loop above this one advances the iterator to a VMA whose `vmStart_ >= startAddr` if `startAddr` is provided (i.e not 0) then breaks so that this loop can look for space between existing VMAs

Reply:
The loop above this one only triggers if a non-zero start address is provided and is only used to advance the iterator

Reply:
So it adds the vma list entry but for the case of a non-zero startAddr performs a redundant check of `if (next->vmStart_ - itr->vmEnd_ >= size)` as that would have already been performed in `if (itr->vmEnd_ <= startAddr && next->vmStart_ >= (startAddr + size)) {`. If I understand you correctly.